libraries/chain/peer_keys_db.cpp

```diff
@@ -0,0 +1,80 @@
+#include <eosio/chain/peer_keys_db.hpp>
+#include <eosio/chain/contract_table_objects.hpp>
+
+namespace eosio::chain {
+
+peer_keys_db_t::peer_keys_db_t() {}
+
+std::optional<public_key_type> peer_keys_db_t::get_peer_key(name n) const {
+   std::lock_guard g(_m);
+   if (auto it = _peer_key_map.find(n); it != _peer_key_map.end())
+      return std::optional<public_key_type>(it->second);
+   return std::optional<public_key_type>{};
+}
+
+size_t peer_keys_db_t::size() const {
+   std::lock_guard g(_m);
+   return _peer_key_map.size();
+}
+
+// we update the keys that were registered up to lib_number (inclusive)
+// --------------------------------------------------------------------
+size_t peer_keys_db_t::update_peer_keys(const controller& chain, uint32_t lib_number) {
+   size_t num_updated = 0;
+   try {
+      if (lib_number <= _block_num)
+         return num_updated;                      // nothing to do
+
+      const auto& db = chain.db();
+      const auto  table_ref = boost::make_tuple("eosio"_n, "eosio"_n, "peerkeys"_n);
+      const auto* t_id      = db.find<table_id_object, by_code_scope_table>(table_ref);
+      EOS_ASSERT(t_id != nullptr, plugin_exception, "cannot retrieve `peerkeys` table");
+
+      const auto& secidx = db.get_index<index64_index, by_secondary>();
+
+      const auto lower = secidx.lower_bound(std::make_tuple(t_id->id._id, static_cast<uint64_t>(_block_num + 1)));
+      const auto upper = secidx.upper_bound(std::make_tuple(t_id->id._id, static_cast<uint64_t>(lib_number)));
+
+      if (upper == lower) {
+         // no new keys registered
+         _block_num = lib_number;
+         return num_updated;
+      }
+
+      std::lock_guard g(_m); // we only need to protect access to _peer_key_map
```

Comment:
I thought this had to be a `fc::lock_guard`.

Reply:
https://github.com/AntelopeIO/spring/pull/1228/commits/985ec9b02226288692ec98e9decaf4eb22f3d941